include/El/core/SimpleBuffer.hpp

```diff
@@ -0,0 +1,125 @@
+#ifndef EL_CORE_SIMPLEBUFFER_HPP_
+#define EL_CORE_SIMPLEBUFFER_HPP_
+
+namespace El
+{
+
+// A simple data management class for temporary contiguous memory blocks
+template <typename T, Device D>
+class simple_buffer
+{
+public:
+    simple_buffer() = default;
+    explicit simple_buffer(size_t size,
+                           unsigned int mode = DefaultMemoryMode<D>());
+    explicit simple_buffer(size_t size, T const& value,
+                           unsigned int mode = DefaultMemoryMode<D>());
+
+    void allocate(size_t size);
+
+    void shallowCopyIfPossible(simple_buffer<T,D>& other)
+    {
+        data_ = other.data();
+        size_ = other.size();
+    }
+
+    template <Device D2>
+    void shallowCopyIfPossible(simple_buffer<T,D2>& other)
+    {
+        data_ = mem_.Require(other.size());
+        size_ = other.size();
+
+        InterDeviceCopy<D2,D>::MemCopy1D(data_, other.data(), size_);
+    }
+
+    size_t size() const noexcept;
+    T* data() noexcept;
+    T const* data() const noexcept;
+
+private:
+    Memory<T,D> mem_;
+    T* data_;
+    size_t size_;
+}; // class simple_buffer
+
+
+namespace details
+{
+template <Device D> struct MemorySetter;
+
+template <>
+struct MemorySetter<Device::CPU>
+{
+    template <typename T>
+    static void setBufferToValue(T* buffer, size_t size, T const& value)
+    {
+        std::fill_n(buffer, size, value);
+    }
+};// struct MemorySetter<Device::CPU>
+
+#ifdef HYDROGEN_HAVE_CUDA
+template <>
+struct MemorySetter<Device::GPU>
+{
+    template<typename T>
+    static void setBufferToValue(T* buffer, size_t size, T const& value)
+    {
+        if( value == T(0) )
```

Comment:
We could probably use the `Fill_GPU_impl()` in src/blas_like/level1/GPU/Fill.cu to do this entirely on GPU.